### ISSUE TITLE
Throwing in datasources causes urania to hang

### ISSUE THREAD

**Issue by lennartbuit** [funcool/urania#16]
Heya :wave:

Datasources are supposed to return a promise that can be resolved or rejected. So, if you have a datasource that rejects, `u/execute!` can work as expected:

```clojure
(defrecord Fail [nonce]
  u/DataSource
  (-identity [_] nonce)
  (-fetch [_ _] (prom/rejected (ex-info "Not good" {}))))
```

```clojure
(-> (u/execute! (Fail. 123))
    (prom/catch (fn [ex] :exception))
    (deref))
=> :exception
```

If you, however, don't return a promise, but throw from your fetch function, the promise returned by `u/execute!` never resolves:

```clojure
(defrecord FailThrow [nonce]
  u/DataSource
  (-identity [_] nonce)
  (-fetch [_ _] (throw (ex-info "Not good" {}))))
```

```clojure
(-> (u/execute! (FailThrow. 123))
    (prom/catch (fn [ex] :exception))
    (deref)) ;; Hangs
```

The 'issue' here is that run-fetch creates a _new_ promise, that is resolved/rejected based on a promise that u/-fetch may return:

https://github.com/funcool/urania/blob/3d3c61206d8a7de675af5d97ad08706ad3307a15/src/urania/core.cljc#L168-L174

So, if fetch throws, then neither the `prom/then` nor the `prom/catch` are executed. That anonymous fn is just interrupted.

-----

So, I can see an argument for _not_ fixing this. The fetch function kinda says that you should return a promise, and there is a way to reject promises. So, maybe my `FailThrow` datasource is not actually adhering to its contract (rejcted promises, instead of throwing).

That said, for me, this generally happens when I make a mistake. If I program something that doesn't work and throws like an NPE, then my dataloader starts hanging...

Thoughts?

**Comment by oliyh** [comment]
Hello,

Yes, this has caught me out in the last as well - the contract for a Urania data source is a promise, it's not allowed to throw exceptions.

That doesn't mean superlifter can't help you, though. Debugging a hanging Lacinia query is never going to be easy, so I think superlifter can be improved to catch these and fail quickly.

Cheers

**Comment by oliyh** [comment]
I don't have anything against it, I think this is a great improvement to the API because it removes a confusing and difficult to trace behaviour. Happy to accept the PR, thanks.

**Comment by oliyh** [comment]
Fixed with #17
